problem/0005-protocol-detection.md

```diff
@@ -0,0 +1,189 @@
+# Protocol Detection Configuration
+
+- Contribution Name: Protocol Detection Configuration
+- Implementation Owner:
+- Start Date: 2020-05-04
+- Target Date:
+- RFC PR:
+- Linkerd Issue:
+- Reviewers: @olix0r @adleong
+
+## Summary
+
+[summary]: #summary
+
+Instead of requiring users to configure the protocol for every port being used
+in an application, Linkerd tries to detect the protocol being used. For
+protocols that the proxy knows about (HTTP) and client speaks first protocols,
+this works pretty well. Unfortunately, for server speaks first protocols, the
+proxy is unable to detect what protocol is being used. This causes delays in
+negotiating new connections and in some cases surprising failure behavior when
+adding Linkerd to new services.
+
+Today, when an application is using an unsupported protocol, the solution is to
+skip either a specific inbound/outbound port or an entire range. When a user
+skips ports on a per-pod basis, the `iptables` rules that redirect incoming and
+outgoing traffic are changed slightly. Any traffic for these skipped ports
+actually skips the proxy. This traffic cannot be mTLS'd and there are no metrics
+available.
+
+## Problem Statement
+
+[problem-statement]: #problem-statement
+
+### What problem are you trying to solve?
+
+Users should be able to get the benefit of lower level protocols (L4) when
+Linkerd is unable to detect the higher level protocol (L7). This allows users to
+get all the benefits they would expect from mTLS, TCP metrics and multi-cluster
+routing. To do this, users need to specify whether protocol detection should be
+skipped on a per-port basis.
+
+### Requirements
+
+- Allow ports to skip the proxy entirely (inbound and outbound), maintaining
+  current behavior.
+- Allow ports to skip only protocol detection inside the proxy (inbound and
+  outbound).
+- Support port ranges as well as individual ports.
+- Surface metrics at the TCP level.
+- Work for applications where both the client and server are meshed.
+- Do not require configuration on both the client and server.
+
+### Out of Scope
+
+- Configuration of the protocol on a port.
```

Comment:
What does this mean?

Reply:
Calling a port `http`, `grpc` or `kafka`. I originally called this in scope and realized that I didn't have good enough of a use case in mind to make it all fit together.